pkg/controllers/ragengine_gc_finalizer.go

```diff
@@ -0,0 +1,64 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+package controllers
+
+import (
+	"context"
+
+	kaitov1alpha1 "github.com/azure/kaito/api/v1alpha1"
+	"github.com/azure/kaito/pkg/featuregates"
+	"github.com/azure/kaito/pkg/machine"
+	"github.com/azure/kaito/pkg/nodeclaim"
+	"github.com/azure/kaito/pkg/utils/consts"
+	"k8s.io/klog/v2"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+)
+
+// garbageCollectRAGEngine remove finalizer associated with ragengine object.
+func (c *RAGEngineReconciler) garbageCollectRAGEngine(ctx context.Context, ragEngineObj *kaitov1alpha1.RAGEngine) (ctrl.Result, error) {
+	klog.InfoS("garbageCollectRAGEngine", "ragengine", klog.KObj(ragEngineObj))
+
+	// Check if there are any machines associated with this ragengine.
+	mList, err := machine.ListMachines(ctx, ragEngineObj, c.Client)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	// We should delete all the machines that are created by this ragengine
+	for i := range mList.Items {
+		if deleteErr := c.Delete(ctx, &mList.Items[i], &client.DeleteOptions{}); deleteErr != nil {
+			klog.ErrorS(deleteErr, "failed to delete the machine", "machine", klog.KObj(&mList.Items[i]))
+			return ctrl.Result{}, deleteErr
+		}
+	}
+
+	if featuregates.FeatureGates[consts.FeatureFlagKarpenter] {
+		// Check if there are any nodeClaims associated with this ragengine.
+		ncList, err := nodeclaim.ListNodeClaim(ctx, ragEngineObj, c.Client)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+
+		// We should delete all the nodeClaims that are created by this ragengine
+		for i := range ncList.Items {
+			if deleteErr := c.Delete(ctx, &ncList.Items[i], &client.DeleteOptions{}); deleteErr != nil {
+				klog.ErrorS(deleteErr, "failed to delete the nodeClaim", "nodeClaim", klog.KObj(&ncList.Items[i]))
+				return ctrl.Result{}, deleteErr
+			}
+		}
+	}
+
+	staleWObj := ragEngineObj.DeepCopy()
+	staleWObj.SetFinalizers(nil)
```

Comment:
we should only remove the finalizer which is set by this controller, or we may remove the finalizer set by user by accident.

Reply:
I think from 53-62 all about removing the finalizer?

Reply:
Suppose finalizers is ["kaito/a", "from-user/b"], just remove the finalizer set by this controller (kaito/b), and set finalizers to ["from-user/b"].
you can take a look at this code https://github.com/openkruise/kruise/blob/master/pkg/controller/ephemeraljob/ephemeraljob_utils.go#L256